unified-runtime/source/adapters/level_zero/v2/queue_immediate_in_order.cpp

```diff
@@ -705,31 +707,153 @@ ur_result_t ur_queue_immediate_in_order_t::enqueueWriteHostPipe(
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY(
+      "ur_queue_immediate_in_order_t::enqueueUSMSharedAllocExp");
+
+  return enqueueUSMAllocHelper(pPool, size, pProperties, numEventsInWaitList,
+                               phEventWaitList, ppMem, phEvent,
+                               UR_USM_TYPE_SHARED);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY("ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp");
+
+  return enqueueUSMAllocHelper(pPool, size, pProperties, numEventsInWaitList,
+                               phEventWaitList, ppMem, phEvent,
+                               UR_USM_TYPE_HOST);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMFreeExp(
-    ur_usm_pool_handle_t, void *, uint32_t, const ur_event_handle_t *,
-    ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t, void *pMem, uint32_t numEventsInWaitList,
+    const ur_event_handle_t *phEventWaitList, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY("ur_queue_immediate_in_order_t::enqueueUSMFreeExp");
+  auto commandListLocked = commandListManager.lock();
+
+  auto zeSignalEvent = getSignalEvent(commandListLocked, phEvent,
+                                      UR_COMMAND_ENQUEUE_USM_FREE_EXP);
+  auto [pWaitEvents, numWaitEvents] =
+      getWaitListView(commandListLocked, phEventWaitList, numEventsInWaitList);
+
+  umf_memory_pool_handle_t hPool = umfPoolByPtr(pMem);
+  if (!hPool) {
+    return UR_RESULT_ERROR_INVALID_MEM_OBJECT;
+  }
+
+  UsmPool *usmPool = nullptr;
+  auto ret = umfPoolGetTag(hPool, (void **)&usmPool);
+  if (ret != UMF_RESULT_SUCCESS || !usmPool) {
+    // This should never happen
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  size_t size = umfPoolMallocUsableSize(hPool, pMem);
+  ur_event_handle_t poolEvent = nullptr;
+  if (phEvent) {
+    poolEvent = *phEvent;
+    poolEvent->RefCount.increment();
+  }
+
+  if (numWaitEvents > 0) {
+    ZE2UR_CALL(
+        zeCommandListAppendWaitOnEvents,
+        (commandListLocked->getZeCommandList(), numWaitEvents, pWaitEvents));
+  }
+
+  if (zeSignalEvent) {
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
+               (commandListLocked->getZeCommandList(), zeSignalEvent));
+  }
+
+  // Insert must be done after the signal event is appended.
+  usmPool->asyncPool.insert(pMem, size, poolEvent, this);
```

Comment:
poolEvent cannot be null, even if phEvent is null.

Otherwise this memory will be immediately available for reuse, before the events on the wait list signal (or, if wait list is empty, before whatever is prior on the queue is completed).

Reply:
Good point, thanks.